### PR TITLE
Add support AmdSev build of OVMF

### PR DESCRIPTION
**Pull request by Alex-Carter01**
Fixes: #4169
Signed-off-by: Alex Carter <alex.carter@ibm.com>

Introduces a build script for AmdSev build of OVMF, which will be needed for SEV measured direct boot.

### CODE REVIEW

File: versions.yaml
Comment (fidencio) [inline versions.yaml:246]: I'd like to ask you to add an subsection called sev under the newly introduced OVMF.  QEMU with TDX also uses OVMF and we can leverage this very same entry later on.

An example of how can be done is: https://github.com/kata-containers/kata-containers/blob/3f069c7acbfa3bd789255ca72f5b56414f4123e3/versions.yaml#L153-L160

Comment (Alex-Carter01) [reply versions.yaml:246]: Sounds good, I have added a new SEV subsection. Glad to know this can be used for tdx in the future.

**Comment by Alex-Carter01** [comment]
> Also, I'd suggest the commit message to be changed to something like:
> 
> ```
> static-build: Add AmdSev build of OMVF
> 
> Introduce a build script for AmdSev build of OMVF, which is needed for
> SEV measured direct boot.
> 
> Fixes: #4168
> Signed-off-by: Alex Carter <alex.carter@ibm.com>
> ```

Updated the commit message to be more informative. Thank you for all the feedback!

**Comment by fidencio** [comment]
@Alex-Carter01, the commit check is failing and the author's email of the committer and the "signed-off-by" do not match.
Please, update that.

@amshinde, mind to also take a look at this as QEMU also uses OVMF for TDX?

**Comment by Alex-Carter01** [comment]
I have added my IBM email to this account, and amended the commit with that email is the author so it matches the signature, I hope that resolves things.

**Comment by fidencio** [comment]
LGTM, but I'd like to hear from folks working on the QEMU side at Intel, so I'll wait for @amshinde to give it a review as well.

**Comment by Alex-Carter01** [comment]
resolved small new conflict in versions file

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (dubek) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:56]: Suggestion: extract the platform file name `OvmfPkg/AmdSev/AmdSevX64.dsc` to a configuration entry as well, and then `build-ovmf.sh` can be used for other OVMF targets (TDX?). However, I see that further down we look for `Build/AmdSev/DEBUG_GCC5/FV/OVMF.fd`, which is AmdSev-specific, so maybe some more config entries are needed? Not sure how generic this should be.

Comment (Alex-Carter01) [reply tools/packaging/static-build/ovmf/build-ovmf.sh:56]: Sure, I can look into what I can do to make this more generic now that I know it will likely also be used for TDX.

Comment (Alex-Carter01) [reply tools/packaging/static-build/ovmf/build-ovmf.sh:56]: Sure, I can look into what I can do to make this more generic now that I know it will likely also be used for TDX.

**Review by amshinde — APPROVED** [review]
lgtm

**Comment by Alex-Carter01** [comment]
Made some changes based on the suggestions from @dubek to make this script more generic for other builds of OVMF. 

The way I would imagine say TDX using this script would be to add another entry in the versions file, now with the desired package, and set a environment variable "ovmf_build" to 'tdx'. I have currently set the default value for this to sev. Then there would need to be a much simpler PR against this to add cases for tdx to change where to pull versions information from, but the rest of the build process should then be flexible.

Thank you @amshinde for the review! Letting you know this push to make it more generic is relatively large compared to past tweaks.

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (larrydewey) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:70]: Same as on line 42, quoting to prevent interpreter problems from occurring.

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (larrydewey) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:85]: Same as 42.

**Review by dubek — APPROVED** [review]

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (wainersm) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:28]: `git clone https://github.com/tianocore/edk2` is enough to clone the repo, you don't need to append `.git` in the url. So this block can be simplified.

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (wainersm) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:42]: Even if `${ovmf_dir}` exists it is not guarantee to be the same sources as from `${ovmf_repo}`. My suggestion is to delete and then clone. Actually, I think it would be better to use a temporary directory for the build (and properly clean up it afterwards), for example:

```
....
build_root=$(mktemp -d)
cleanup() {
    if [ "${DEBUG:-}" == true ]; then
       info "Not deleted build root directory: ${build_root}"
       return
    fi
    rm -rf "${build_root}"
}
....
trap cleanup EXIT
```

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (wainersm) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:66]: Move this (and the other initializations below) up at the begin of the script so that if some information for the build is missing then it exists right away (fail fast).

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (wainersm) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:52]: s/scipt/script/

**Comment by wainersm** [comment]
Hi @Alex-Carter01 , thanks for working on this. I finished my review, left just a couple of comments.

**Comment by Alex-Carter01** [comment]
This change accounts for 3/5 of @wainersm 's suggestions. I will be coming back to add a fix to accommodate DESTDIR/PREFIX values and a temporary build directory

**Comment by wainersm** [comment]
> This change accounts for 3/5 of @wainersm 's suggestions. I will be coming back to add a fix to accommodate DESTDIR/PREFIX values and a temporary build directory

Great. thanks @Alex-Carter01

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (fidencio) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:21]: The other scripts are using `${PWD}` as the destdir, any specific reason to use `../../destir`?

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (fidencio) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:23]: The default build should be the "neutral" one, the `x64` one.  Then in the kata-deploy scripts we pass whether we'll build OVMF for SEV or TDVF for TDX.

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (fidencio) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:42]: Extra empty line.

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (fidencio) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:43]: I'd prefer to have those being introduced in two different commits.
The first one adding support for building OVMF, and a second one adding support for building OVMF for SEV.  That would be cleaner.

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (fidencio) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:41]: I know this is how it's called on OVMF side, but for consistency sake I'd prefer having `x64` renamed to `x86_64` both here and in the versions file.

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (fidencio) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:94]: I wouldn't do the cleanup.  Leave the directory there, always, as already done for different build scripts.

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (fidencio) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:74]: Is this a leftover to show the directory you're at?  I think this can be removed.

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (fidencio) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:84]: What are the files you're coping?
No need to copy `OVMF_{CODE,VARS}.fd`? If not, why not? :-)

**Review by fidencio — COMMENTED** [review]
@Alex-Carter01, This one looks good, and I'm making several suggestions to what I think that would make it cleaner.

Please, let me know your thoughts on those.

**Comment by fidencio** [comment]
One last bit, it'd be good if we can have it already being built in a container.  So, I'd suggest having a dockerfile introduced as part of the first commit, in a similar way we already have it for QEMU / Kernel.

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (fidencio) [inline tools/packaging/static-build/ovmf/build-ovmf.sh:26]: @Alex-Carter01, what about not hardcoding this ones?
We could default to those values, but let the person doing the build specify env-vars for those, as done with the ones above.

**Comment by Alex-Carter01** [comment]
These changes are for most of the feedback from @fidencio regarding having the first commit be for generic (X86_64) OVMF script. This commit still needs an accompanying dockerfile.

The second commit will reinstate the AmdSev build options

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (Alex-Carter01) [reply tools/packaging/static-build/ovmf/build-ovmf.sh]: Variables now allow specifying different values

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (Alex-Carter01) [reply tools/packaging/static-build/ovmf/build-ovmf.sh]: The content of that directory should just be the OVMF.fd file

File: tools/packaging/static-build/ovmf/build-ovmf.sh
Comment (Alex-Carter01) [reply tools/packaging/static-build/ovmf/build-ovmf.sh]: Renamed, and made the default option. It is the only inclusion in the first commit.

**Comment by fidencio** [comment]
@Alex-Carter01, the first commit seems good from a quick look.
I'd like to have the dockerfile as part of this first commit, and then the SEV part as part of a second commit, in this very same PR.

Makes sense?

**Comment by Alex-Carter01** [comment]
The first commit now has the dockerfile. The new build script makes the docker container and then runs the ovmf script inside of it. My next step is adding the second commit with the AmdSev additions.

**Comment by Alex-Carter01** [comment]
Added the second commit with the extra option for the AmdSev build

File: tools/packaging/static-build/ovmf/build.sh
Comment (fidencio) [inline tools/packaging/static-build/ovmf/build.sh:23]: As mentioned above, I'd add more environment variables to be passed / respected here, so `yq` can be removed as a hard dependency.

**Review by fidencio — CHANGES_REQUESTED** [review]
@Alex-Carter01, thanks for the work on this, I've added a few suggestions.

File: tools/packaging/static-build/ovmf/Dockerfile
Comment (Alex-Carter01) [reply tools/packaging/static-build/ovmf/Dockerfile:21]: I found it interesting, but yes the build fails without both python and python3

File: tools/packaging/static-build/ovmf/build.sh
Comment (Alex-Carter01) [reply tools/packaging/static-build/ovmf/build.sh:23]: I will cleanup the ordering and use variables outside so yq isn't needed in the docker container.

**Comment by Alex-Carter01** [comment]
This change passes variables into the docker environment so yq is not needed inside the container.

File: tools/packaging/static-build/ovmf/Dockerfile
Comment (fidencio) [reply tools/packaging/static-build/ovmf/Dockerfile:20]: Yeah, weird. :-)

File: tools/packaging/static-build/ovmf/Dockerfile
Comment (fidencio) [inline tools/packaging/static-build/ovmf/Dockerfile:7]: As `yq` is not installed anymore, this bit should be dropped.

And once it's dropped, we're good to have this one merged.  Thanks @Alex-Carter01!

**Comment by Alex-Carter01** [comment]
This change removes the holdover go path variable now that yq is no longer needed inside the docker container from the first commit.

**Comment by Alex-Carter01** [comment]
Fixing a static check I saw failing for trailing whitespace in the versions file.

**Comment by wainersm** [comment]
Hi @Alex-Carter01 @fidencio ,  I cloned this on my local Fedora 34 x86_64 machine, ran the build script and it built OVMF just fine. I didn't check the built firmware works as I don't have a SEV machine. Anyway, I think it is ready to land in main...so we keep the work to enable SEV CI.

**Review by fidencio — APPROVED** [review]
lgtm, thanks @Alex-Carter01!

**Comment by fidencio** [comment]
> Hi @Alex-Carter01 @fidencio , I cloned this on my local Fedora 34 x86_64 machine, ran the build script and it built OVMF just fine. I didn't check the built firmware works as I don't have a SEV machine. Anyway, I think it is ready to land in main...so we keep the work to enable SEV CI.

Glad your comment was positive, as it showed up just after I merged the PR. :-)

**Pull request merged**
